File: js/permutations.js
function printPermutations(a, n) {
  if (n === 0) {
    console.log(a.join("\t"))
  } else {
    for (let i = 0; i < n; i++) {
      printPermutations(a, n - 1)
      const j = n % 2 === 0 ? 0 : i
      ;[a[j], a[n]] = [a[n], a[j]]
    }
    printPermutations(a, n - 1)
  }
}

const args = process.argv.slice(2)
printPermutations(args, args.length - 1)

File: perl/permutations.pl
use v5.40;

sub print_permutations($a, $n) {
    if ($n <= 0) {
        print join("\t", @$a), "\n";
    } else {
        for my $i (0 .. $n - 1) {
            print_permutations($a, $n - 1);
            my $j = $n % 2 == 0 ? 0 : $i;
            @$a[$j, $n] = @$a[$n, $j];
        }
        print_permutations($a, $n - 1);
    }
}

print_permutations(\@ARGV, scalar(@ARGV) - 1);
